app/assets/javascripts/discourse/widgets/dropdown.js

```diff
@@ -0,0 +1,266 @@
+import { createWidget } from "discourse/widgets/widget";
+import hbs from "discourse/widgets/hbs-compiler";
+
+/*
+
+  dropdown widget
+
+  Usage
+  -----
+
+  {{attach
+    widget="dropdown"
+    attrs=(hash
+      id=id
+      label=label
+      content=content
+      onChange=onChange
+      options=(hash)
+    )
+  }}
+
+  Mandatory attributes:
+
+    - id: must be unique in the application
+
+    - label: an i18n key to be translated and displayed on the header
+
+    - translatedLabel: an already translated label to display on the header
+
+    - onChange: action called when a click happens on a row, content[rowIndex] will be passed as params
+
+  Optional attributes:
+
+    - class: adds css class to the dropdown
+    - content: list of items to display, if undefined or empty dropdown won't display
+      Example content:
+
+      ```
+      [
+        { id: 1, label: "foo.bar" },
+        "separator",
+        { id: 2, translatedLabel: "FooBar" },
+        { id: 3 label: "foo.baz", icon: "times" }
+      ]
+      ```
+
+    - options: accepts a hash of optional attributes
+      - smallHeader: true/false, if true "btn-small" class will be added to header.
+*/
+
+export const DropdownHeaderClass = {
+  tagName: "button",
+
+  transform(attrs) {
+    return {
+      label: attrs.translatedLabel ? attrs.translatedLabel : I18n.t(attrs.label)
+    };
+  },
+
+  buildClasses(attrs) {
+    const classes = ["widget-dropdown-header", "btn", "btn-default"];
+
+    if (attrs.options.small) {
+      classes.push("btn-small");
+    }
+
+    return classes.join(" ");
+  },
+
+  click(event) {
+    event.preventDefault();
+
+    this.sendWidgetAction("_onTrigger");
+  },
+
+  template: hbs`
+    {{#if attrs.icon}}
+      {{d-icon attrs.icon}}
+    {{/if}}
+    <span class="label">
+      {{transformed.label}}
+    </span>
+  `
+};
+
+createWidget("dropdown-header", DropdownHeaderClass);
+
+export const DropdownItemClass = {
+  tagName: "div",
+
+  transform(attrs) {
+    return {
+      content:
+        attrs.item === "separator"
```

Comment:
I can't decide if I hate this or not :smiley: 

On one hand nested ternary operators make my brain hurt, but on the other I fully understand what this is doing. I am leaning towards "let's keep it."

Reply:
Yes I feel the same :D